src/core/ext/transport/inproc/inproc_transport.cc

```diff
@@ -847,14 +865,13 @@ void op_state_machine_locked(inproc_stream* s, grpc_error* error) {
         "op_state_machine scheduling recv-message-on-complete");
     s->recv_message_op = nullptr;
   }
-  if (s->trailing_md_recvd && (s->trailing_md_sent || s->t->is_client) &&
-      s->send_message_op) {
+  if (s->trailing_md_recvd && s->send_message_op && s->t->is_client) {
```

Comment:
with this change -
1) the server will try to send messages even after it has sent trailing metadata..well, this case should not happen in the first place, so it would be fine
2) If the client has already sent trailing md, it will still try to send messages but again this case should not happen.

This change should be fine but I don't understand what this is solving

Reply:
1. The inproc transport is a little funny since it has 0 flow control for messages. As a result, a send doesn't go out until it is matched to a receive. In this case, the status can actually go out to the client before the last message if the client has already requested status. I do think that that is a problem and will be tracked in #23398. I would even call that a blocking issue for de-experimentalizing the callback API. (It's a non-issue for the sync and async API since those APIs won't have the client request status until it won't ask for any more receive messages.) My plan is to always have 1 flow control message from server-to-client and then block the actual send status until the send message has gone to flow control. But in the meanwhile, this change is sufficient to stop the actual flakes that have been arising.

2. On the client side, this doesn't change anything. The purpose of the following block is for the client to consider that any send that it might have to send won't get matched to any further receive if the server has already sent status.

Reply:
I see